pkg/apis/pubsub/v1alpha1/topic_types.go

```diff
@@ -0,0 +1,160 @@
+ */
+
+package v1alpha1
+
+import (
+	"github.com/knative/pkg/apis"
+	"github.com/knative/pkg/apis/duck"
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	duckv1beta1 "github.com/knative/pkg/apis/duck/v1beta1"
+	"github.com/knative/pkg/webhook"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Topic is a resource representing a Topic backed by Google Cloud Pub/Sub.
+type Topic struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the Topic.
+	Spec TopicSpec `json:"spec,omitempty"`
+
+	// Status represents the current state of the Topic. This data may be out of
+	// date.
+	// +optional
+	Status TopicStatus `json:"status,omitempty"`
+}
+
+// Check that Topic can be validated, can be defaulted, and has immutable fields.
+var _ runtime.Object = (*Topic)(nil)
+var _ webhook.GenericCRD = (*Topic)(nil)
+
+// Check that Topic implements the Conditions duck type.
+var _ = duck.VerifyType(&Topic{}, &duckv1beta1.Conditions{})
+
+// TopicSpec defines an parameters for creating or publishing to a Cloud
+// Pub/Sub Topic depending the PropagationPolicy.
+type TopicSpec struct {
+	// Secret is the credential to be used to create and publish into the
+	// Cloud Pub/Sub Topic. The value of the secret entry must be a service
+	// account key in the JSON format
+	// (see https://cloud.google.com/iam/docs/creating-managing-service-account-keys).
+	Secret *corev1.SecretKeySelector `json:"secret,omitempty"`
+
+	// Project is the ID of the Google Cloud Project that the Pub/Sub
+	// Topic will be created in or used from.
+	Project string `json:"project,omitempty"`
+
+	// Topic is the ID of the Topic to create/use in Google Cloud Pub/Sub.
+	Topic string `json:"topic,omitempty"`
+
+	//PropagationPolicy defines how Topic controls the Cloud Pub/Sub topic for lifecycle changes.
+	PropagationPolicy PropagationPolicyType `json:"propagationPolicy,omitempty"`
+
+	// ServiceAccoutName is the name of the ServiceAccount that will be used to
+	// run the Topic Deployment.
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+}
+
+// PropagationPolicyType defines enum type for TopicPolicy
+type PropagationPolicyType string
+
+const (
+	// TopicPolicyCreateDelete defines the Cloud Pub/Sub topic management
+	// policy for creating topic if not present, and delete topic when the
+	// Topic resource is deleted.
+	TopicPolicyCreateDelete PropagationPolicyType = "CreateDelete"
+
+	// TopicPolicyCreateRestrictDelete defines the Cloud Pub/Sub topic
+	// management policy for creating topic if not present, and do not delete
+	// topic when the Topic resource is deleted.
+	TopicPolicyCreateRestrictDelete PropagationPolicyType = "CreateRestrictDelete"
```

Comment:
Restrict is a bit confusing for me because it implies there is a range of possible restrictions that could be applied. It's harder to determine what the behavior will be. I suggest renaming to something like `CreateNoDelete` or `CreateNeverDelete`.